lm_eval/tasks/qasper.py

```diff
@@ -0,0 +1,222 @@
+            "TITLE: "
+            + doc["title"]
+            + "\n"
+            + "ABSTRACT: "
+            + doc["abstract"]
+            + "\n\n"
+            + "Q: "
+            + doc["question"]
+            + "\n\n"
+            + "A:"
+        )
+
+    def doc_to_target(self, doc):
+        # this method is invoked by tests only
+        answer = doc["answer"]
+        if isinstance(answer, list):
+            answer = ", ".join(answer)
+        return " " + answer
+
+    def training_docs(self):
+        for doc in self.data["train"]:
+            yield from self.process_doc(doc)
+
+    def validation_docs(self):
+        for doc in self.data["train"]:
+            yield from self.process_doc(doc)
+
+    def process_doc(self, doc):
+        """Given a `doc`, flatten it out so that each JSON blob
+        contains exactly one question and one answer. Logic taken from
+        the reference implementation available at
+        https://github.com/allenai/qasper-led-baseline/blob/main/scripts/evaluator.py
+        """
+        obs_list = []
+        for question, answer_list in zip(doc["qas"]["question"], doc["qas"]["answers"]):
+            for answer_blob in answer_list["answer"]:
+                answer, answer_type = categorise_answer(answer_blob)
+                obs_list.append(
+                    {
+                        "title": doc["title"],
+                        "abstract": doc["abstract"],
+                        "question": question,
+                        "answer": answer,
+                        "answer_type": answer_type,
+                    }
+                )
+        return obs_list
+
+    def process_results(self, doc, results):
+        # TODO: Calculate a score for extractive spans once a request type for generating
+        # extractive spans is available
+        if len(results) == 1:
+            [(logprob_unanswerable, _)] = results
+        elif len(results) == 2:
+            res, (logprob_unanswerable, _) = results
+        else:
+            ll_yes, ll_no, (logprob_unanswerable, _) = results
+        res_dict = {}
+
+        # Handle unanswerability first
+        unanswerable_gold = doc["answer_type"] == "unanswerable"
+        unanswerable_pred = exp(logprob_unanswerable) > 1 - exp(logprob_unanswerable)
+        res_dict["f1_un"] = (unanswerable_gold, unanswerable_pred)
+
+        # Handle yes/no questions
+        if doc["answer_type"] == "bool":
+            gold = 1 if doc["answer"] == "yes" else 0
+            pred = ll_yes > ll_no
+            res_dict["f1_yn"] = (gold, pred)
+
+        # Handle completions
+        if doc["answer_type"] == "free form answer":
+            res_dict["f1_ab"] = token_f1_score(res, doc["answer"])
+
+        # Handle extraction
+        if doc["answer_type"] == "extractive_spans":
+            res_dict["f1_ex"] = 0
+        return res_dict
+
+    def aggregation(self):
+        return {"f1_un": f1_score, "f1_yn": f1_score, "f1_ab": mean, "f1_ex": mean}
```

Comment:
double checking, are the latter 2 supposed to be mean, and is yesno supposed to be f1?